### DISCUSSION TITLE
is there a 49.7 day millis() overflow in the code?

### DISCUSSION THREAD

**Discussion by peff74** [hoylabs/OpenDTU-OnBattery#1224] [Q&A]
I noticed again yesterday that the OpenDTUonBatt stops working after 49.7 days.
Only PowermeterClass can still be seen in the logs.
![image](https://github.com/user-attachments/assets/7ea4dd10-fbb1-4254-8aee-42b8f61e6b21)
Everything else is then dead, MQTT, the control of the Hoymile, etc.
The Hoymiles then simply continues to run on its last setting... because it no longer receives any commands.

**Comment by schlimmchen** [comment]
There is more than just one of those bugs.

It is absolutely conceivable that such an overflow grinds the whole `loop()` to a halt, which is what you seem to observe.

Congrats on running your OpenDTU-OnBattery for nearly 50 days straight! :partying_face: :rocket: Unfortunately, unless you have the last lines of logs before the issue occured, we will not find out where the problem is, at least not any time soon.

However, seeing that the whole unit stops working, which is significantly worse than "oops, this calculation is wrong once every 50 days", we might need to investigate. We might have a chance to override millis() to 49.69 days and hopefully are able to reproduce the issue fast, then we can start pinning it down.

Could you maybe proof that your uptime is 49.7 days and then create an issue?

**Reply by skippermeister** [reply]
@schlimmchen 
Wenn in den Timerabfragen steht
If ( …. millis() > lastTime + period) dann passiert genau der Fehler nach 49.7 Tagen, also beim 32 Bit Überlauf des internen Timers der Funktion millis().

Wird aber geschrieben

if ( …. millis() - lastTime > period) dann wird der Überlauf in der Subtraktion berücksichtigt. In vielen Timerfunktionen ist das im Sourcecode berücksichtigt, aber leider nicht überall.
